### ISSUE TITLE
Add RQ support

### ISSUE THREAD

**Issue by luongthomas** [DataDog/dd-trace-py#962]
I was wondering if there was any work being done to support [RQ](https://github.com/rq/rq) so that job information is submitted to Datadog.  

If not, then I'm considering implementing the feature myself (but I think I might need a little guidence).

**Comment by brettlangdon** [comment]
hey @luongthomas, I actually ran into someone else recently who was using this.

We do not currently have RQ on our roadmap but if you were interested in writing an integration we'd be happy to try and help out!

**Comment by brettlangdon** [comment]
I definitely don't know enough about `rq`, but from a quick glance at the code, it seems like the majority of what we'd want to trace is in `rq/job.py` in the `Job` class.

**Comment by sandys** [comment]
would like to add to this request. we are moving from celery to rq and would be very grateful for this integration.

**Comment by sburns** [comment]
Hi everyone, wanted to share something we've come up with around tracing RQ jobs...

```
from ddtrace.contrib.django.conf import settings
from rq.job import Job

tracer = settings.TRACER  # build a tracer however you'd like, we configure w/ Django so YMMV

JOB_SERVICE_NAME = 'rq-worker'

class DatadogTracedJob(Job):
    def perform(self):
        with tracer.trace(self.func_name, service=JOB_SERVICE_NAME):
            super().perform()
```

Then you should configure RQ to use this custom Job class.

This isn't perfect, however. We're noticing a small portion of spans come through. What I think is happening is that because RQ uses a forking model in which the code above runs in the "horse" process (their words, not mine :) that exits quickly after the job finishes, the internal queue that holds spans and submits to the agent on an interval is destroyed before this submission. We dealt w/ this when [integrating with Elastic's APM](https://discuss.elastic.co/t/python-agent-only-recording-some-background-tasks/149707) and it just smells like the same underlying issue.

@brettlangdon  is there a way to force this queue to flush? If so, a good place to do that would be just after the `with tracer.trace()`context block so we'd know that the spans from the trace have been enqueued but before the job finishes (and the process dies). Alternatively, and this smells to me, would be to `sleep()` for the interval the writer periodically writes so we could ensure a flush to the agent would be triggered before the job finishes.

Obviously this isn't _great_ but depending on average background job length it may or may not have a detrimental impact on overall job processing capacity of your system.

**Comment by brettlangdon** [comment]
hey @sburns great find, and yeah from the sounds of it, I'd guess that is the issue as well (we have the same problem with Resque for Ruby).

You can force a flush with the following:

```python
from ddtrace import tracer
tracer.writer.stop()
tracer.writer.flush()
```

**Comment by sburns** [comment]
> (we have the same problem with Resque for Ruby)

funny, because the design of RQ greatly inherits from resque :)

**Comment by sburns** [comment]
On 0.29.0, what should happen after the `tracer.trace` context block should be:

```python
tracer.writer.write()
tracer.writer._worker.stop()
tracer.writer._worker.flush_queu()
```

**Comment by brettlangdon** [comment]
@sburns you are right, sorry, I was looking at `master` and thought those changes were already released.

Please be sure upon upgrading to `0.30.0` what I posted should work.

**Comment by Kyle-Verhoog** [comment]
@sburns you can set the "resource" of an span in your tracing line like this:

```python
 with tracer.trace(self.func_name, service=JOB_SERVICE_NAME, resource=self.func_name):
```

can then you should see the different jobs listed in that view.

And by the way, we have an `rq` integration in the works coming very soon 😄

**Comment by rousell** [comment]
We were able to get this to work by doing something along the lines of:
`with tracer.trace("service_name", service=JOB_SERVICE_NAME, resource.self.func_name)`
which then put all of the operations for the service under the same "service_name", and which allowed us to see a list of the resource on the main dashboard for our service. This page in the documentation helped particularly: https://docs.datadoghq.com/tracing/guide/configuring-primary-operation/

**Comment by kylejohnson** [comment]
@Kyle-Verhoog Any update on that `rq` integration?  It would be very useful for a project that I'm currently working on.

**Comment by Kyle-Verhoog** [comment]
Hi @kylejohnson!

Unfortunately some other stuff came up that had precedence. But it's not forgotten. The integration is mostly complete in #1111. We just have to get it over the finish line. I'll see if I can get it back up to date to be reviewed by the team 🙂

**Comment by kylejohnson** [comment]
@Kyle-Verhoog Nice, thank you for the update!  I'm happy to see all of the work that has already been done in #1111.

**Comment by kylejohnson** [comment]
The custom job class that @sburns posted above works pretty well for me, but I'm having some trouble accurately reporting services and spans, and associating child and parent traces.

Could someone give me some tips on how to instrument traces in a case similar to as follows?  Ideally, I'd like for a single trace to be followed from the entry point, through rq, and in to the various worker jobs.  Such as: A single `trace_id` is created for each domain, and follows the job in to `rq` and then to `certifier.verify_domain`.  If this isn't the right place to ask, please let me know.

Below are excerpts of code to show what I'm currently doing:

1. `main.py` which submits jobs to a number of queues.  This script is essentially the entry point.  Called as `ddtrace-run ./main.py`
```
    @tracer.wrap('main', service='DM_SSL')
    def run(self):
       ...

       for domain in domains:

            q_span = tracer.trace('enqueue_domain')
            q_span.set_tag('domain.name', domain)

            verify = self.q_verify.enqueue(
                certifier.verify_domain,
                args=(domain, secret_dm, self.endpoint),
                description=f"Verify {domain}"
            )
            issue = self.q_issue.enqueue(
                certifier.issue_certificate,
                args=(domain, acme_secret, pdns_secret, secret_vault, self.endpoint),
                description=f"Issue SSL certificate for {domain}",
                depends_on=verify,
            )

...

            q_span.finish()
```
2. rq picks up jobs and uses `custom_job_worker.py` to process jobs.  These are ran via supervisor as `ddtrace-run /usr/local/bin/rq worker $queue --url redis://... --job-class 'custom_job_worker.DatadogTracedJob'`
```
from rq.job import Job
from ddtrace import tracer
import time

class DatadogTracedJob(Job):
    def perform(self):
        with tracer.trace('rq-worker', service='DM_SSL', resource=self.func_name):
            super().perform()

        time.sleep(6) # Hack to allow trace writer time to finish writing.

        tracer.writer.write()
        tracer.writer.stop()
        tracer.writer.flush_queue()
```
3. rq calls `certifier.py` to do the actual work.  This file has various functions which perform the actual work, such as for the queue shown above:
```
@tracer.wrap('certifier', service='DM_SSL')
def verify_domain(domain, secret, endpoint):

    try:
        cname = pydig.query(domain, "CNAME")
    except:
        logger.error('Failed looking up CNAME', extra={'domain': f"{domain}", 'step': 'verify_domain', 'status': 'err'})
        raise
    ...

@tracer.wrap('certifier', service='DM_SSL')
def issue_certificate(domain, acme_secret, pdns_secret, secret_vault, endpoint):
    ...
```

**Comment by apenney** [comment]
We're about to switch over to rq (away from celery) and stumbled over the pull request that's open for this.  Is this work likely to get merged any time soon in the current state?

**Comment by Kyle-Verhoog** [comment]
:wave: hey all! Sorry for the delays (I really want to get this in too!). One major blocker on the integration (#1111) was that we didn't have the tooling necessary to get adequate test coverage for the `rq` workers, which I really want to make sure we have before shipping it. I'm hoping that #1546 will make that possible :smile:.


@kylejohnson I missed your questions! Yeah, I think what you're getting at is distributed tracing. You can see how this is done in the soon-to-be integration here: https://github.com/DataDog/dd-trace-py/pull/1111/files#diff-832c956d63673ca9b42e67ed1f5da676R87. You essentially inject tracing context (trace_id, parent_id) into the request and then [extract it](https://github.com/DataDog/dd-trace-py/pull/1111/files#diff-832c956d63673ca9b42e67ed1f5da676R114) on the other end.

Let me know if this answers your question or if you have any troubles setting it up! Sorry for the delay again! 🙂

**Comment by bfitzsimmons** [comment]
Hi there.

How are things coming along on this feature? Given the merge of #1546, is #1111 going to be brought back to life?

We have quite a few projects that need `rq` monitoring, and would love to be able to have visibility into the jobs.

Any information you can provide is appreciated.

**Comment by Kyle-Verhoog** [comment]
@bfitzsimmons yes, thanks for the reminder! I'm going to take a whack at getting it back up to date.

**Comment by snuderl** [comment]
@Kyle-Verhoog Any updates? We are really looking forward to this.

**Comment by miketheman** [comment]
Once `rq` is complete, please consider adding support for [`arq`](https://arq-docs.helpmanual.io/)

**Comment by adamsenn** [comment]
@Kyle-Verhoog another bump ;-) We are using RQ on ~30 projects, and we're looking at moving from NewRelic to DataDog. Built-in support for would be very helpful!

**Comment by brettlangdon** [comment]
This integration is now merged, it'll be released in 0.54.0 no exact eta, but since we just released 0.53.0 it'll probably be ~2 weeks.
